presto-main/src/main/java/com/facebook/presto/tdigest/TDigest.java

```diff
@@ -102,51 +105,49 @@
 
     private final double publicCompression;
     private final double compression;
+    private final int maxCentroidCount;
+    private final int maxBufferSize;
 
     // points to the first unused centroid
     private int activeCentroids;
-
     private double totalWeight;
-
-    private final double[] weight;
-
-    private final double[] mean;
-
-    private double unmergedWeight;
+    private double[] weight;
+    private double[] mean;
 
     // this is the index of the next temporary centroid
     // this is a more Java-like convention than activeCentroids uses
     private int tempUsed;
-    private final double[] tempWeight;
-    private final double[] tempMean;
+    private double unmergedWeight;
+    private double[] tempWeight;
+    private double[] tempMean;
 
     // array used for sorting the temp centroids
     // to avoid allocations during operation
-    private final int[] order;
+    private int[] order;
 
     private TDigest(double compression)
     {
         // ensure compression >= 10
-        // default size = 2 * ceil(compression)
-        // default bufferSize = 5 * size
-        // ensure size > 2 * compression + weightLimitFudge
-        // ensure bufferSize > 2 * size
-
+        // default maxCentroidCount = 2 * ceil(compression)
+        // default maxBufferSize = 5 * maxCentroidCount
+        // ensure maxCentroidCount > 2 * compression + weightLimitFudge
+        // ensure maxBufferSize > 2 * maxCentroidCount
+        checkArgument(!isNaN(compression));
         checkArgument(compression <= MAX_COMPRESSION_FACTOR, "Compression factor cannot exceed %s", MAX_COMPRESSION_FACTOR);
         this.publicCompression = max(compression, 10);
         // publicCompression is how many centroids the user asked for
         // compression is how many we actually keep
         this.compression = 2 * publicCompression;
 
         // having a big buffer is good for speed
-        int bufferSize = 5 * (int) ceil(this.publicCompression + sizeFudge);
-        int size = (int) ceil(this.compression + sizeFudge);
+        maxBufferSize = 5 * (int) ceil(this.publicCompression + sizeFudge);
+        maxCentroidCount = (int) ceil(this.compression + sizeFudge);
 
-        weight = new double[size];
-        mean = new double[size];
-        tempWeight = new double[bufferSize];
-        tempMean = new double[bufferSize];
-        order = new int[bufferSize];
+        weight = new double[4];
```

Comment:
What's the rational behind these numbers? Can you give a high level description of this change?

Reply:
I changed them all to 1 when initializing. While testing with tpch and grouping by different keys, I noticed that many groups were associated with an empty T-Digest. Each t-digest was initializing 2 double arrays of size 230, and 3 double arrays of size 650. Therefore, an empty t-digest was using up about 16_000 bytes of memory, compare to 90 used by quantile digest, which was causing my local heap to blow up when I ran queries with only 300_000+ distinct keys. Actually, this change is associated with another PR #13160 .

Reply:
Please remove the commit from this PR if that's the case.